src/RazorSdk/SourceGenerators/RazorSourceGenerator.cs

```diff
@@ -0,0 +1,280 @@
+            });
+
+            tagHelperFeature.Compilation = GeneratorExecutionContext.Compilation.AddSyntaxTrees(results.Take(files.Count));
+            ArrayPool<SyntaxTree>.Shared.Return(results);
+
+            var lastUpdatedReferenceUtc = GetLastUpdatedReference(GeneratorExecutionContext.Compilation.References);
+            IReadOnlyList<TagHelperDescriptor> refTagHelpers;
+
+            if (lastUpdatedReferenceUtc is not null && lastUpdatedReferenceUtc < File.GetLastWriteTimeUtc(razorContext.RefsTagHelperOutputCachePath))
+            {
+                // Producing tag helpers from a Compilation every time is surprisingly expensive. So we'll use some caching strategies to mitigate this until
+                // we can improve the perf in that area.
+
+                // TagHelpers can come from two locations - the declaration files
+                // and the assemblies participating in the compilation.
+                // In a typical inner loop, the assemblies referenced by the project do not change. We could cache these separately from the tag helpers produced
+                // by the app to avoid some per-compilation costs.
+                // We determine if any of the reference assemblies have a newer timestamp than the output cache for the tag helpers. If not, we can re-use previously
+                // calculated results.
+                refTagHelpers = TagHelperSerializer.Deserialize(razorContext.RefsTagHelperOutputCachePath);
+            }
+            else
+            {
+                tagHelperFeature.DiscoveryMode = TagHelperDiscoveryFilter.ReferenceAssemblies;
+                refTagHelpers = tagHelperFeature.GetDescriptors();
+                TagHelperSerializer.Serialize(razorContext.RefsTagHelperOutputCachePath, refTagHelpers);
+            }
+
+            tagHelperFeature.DiscoveryMode = TagHelperDiscoveryFilter.CurrentCompilation;
+            var assemblyTagHelpers = tagHelperFeature.GetDescriptors();
+
+            var result = new List<TagHelperDescriptor>(refTagHelpers.Count + assemblyTagHelpers.Count);
+            result.AddRange(assemblyTagHelpers);
+            result.AddRange(refTagHelpers);
+
+            return result;
+        }
+
+        private static DateTime? GetLastUpdatedReference(IEnumerable<MetadataReference> references)
+        {
+            DateTime lastWriteTimeUtc = DateTime.MinValue;
+
+            foreach (var reference in references)
+            {
+                // We expect all references in the compilation context to be backed by a file on disk. If not,
+                // we'll bail out and regenerate the tag helper cache where this is invoked.
+                if (reference is not PortableExecutableReference portableExecutableReference || string.IsNullOrEmpty(portableExecutableReference.FilePath))
+                {
+                    return null;
+                }
+
+                var fileWriteTime = File.GetLastWriteTimeUtc(portableExecutableReference.FilePath);
```

Comment:
It's possible I'm misreading the flow here, but are we assuming that the last write time of a `.dll` on disk has some relationship to whether our cache is up-to-date?

What happens when someone changes a reference to an *older* version of a package (which was already cached locally) - don't we need to regenerate the compilation output then?

Reply:
Yup, it's the same issue we've sorted multiple times with the Blazor SDK. We could make this more robust, but we're also waiting for better support for this sort of incrementality from the compiler (tracked https://github.com/dotnet/roslyn/issues/51257).

Reply:
The most important feature of a compiler is correctness. This is violating assumptions we have about disk impact that negate correctness.

Reply:
Directly accessing the file system from a generator violates the assumptions we have around compilation. Performance is important but correctness is more so.

Reply:
> We could make this more robust

Yes, I'd think we need to do so without waiting for new compiler features. I don't know if there was already an issue filed to track this, so just in case I filed https://github.com/dotnet/sdk/issues/15913

Reply:
Ah, now I realise that @captainsafia has already filed https://github.com/dotnet/aspnetcore/issues/30234

Would it make sense to keep https://github.com/dotnet/sdk/issues/15913 open to track the actual correctness issue (which should be addressed in the short term), and https://github.com/dotnet/aspnetcore/issues/30234 to track the broader performance/architectural issue of accessing the filesystem at all (which can wait longer)?

If @captainsafia or @pranavkm prefers there being just one issue for all of this, please feel free to close mine as a duplicate.

Reply:
@SteveSandersonMS Having another issue is fine. If only to capture this particular side-effect of requiring read/write interactions to compute the cache. Ultimately, whatever work is done here will be usurped by what we do long term do deal with the FS-access issue.